rust/src/lockfile.rs

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+/* Copied and adapted from: treefile.rs
+ */
+
+use failure::Fallible;
+use serde_derive::{Deserialize, Serialize};
+use serde_json;
+use std::collections::HashMap;
+use std::path::Path;
+use std::io;
+
+use crate::utils;
+
+/// Parse a JSON lockfile definition.
+fn lockfile_parse_stream<R: io::Read>(input: &mut R,) -> Fallible<LockfileConfig> {
+    let lockfile: LockfileConfig = serde_json::from_reader(input).map_err(|e| {
+        io::Error::new(
+            io::ErrorKind::InvalidInput,
+            format!("serde-json: {}", e.to_string()),
+        )
+    })?;
+    Ok(lockfile)
+}
+
+/// Given a lockfile filename, parse it
+fn lockfile_parse<P: AsRef<Path>>(filename: P,) -> Fallible<LockfileConfig> {
+    let filename = filename.as_ref();
+    let mut f = io::BufReader::new(utils::open_file(filename)?);
+    filename.file_name().map(|s| s.to_string_lossy()).ok_or_else(
+        || io::Error::new(io::ErrorKind::InvalidInput, "Expected a filename"))?;
+    let lf = lockfile_parse_stream(&mut f).map_err(|e| {
+        io::Error::new(
+            io::ErrorKind::InvalidInput,
+            format!("Parsing {}: {}", filename.to_string_lossy(), e.to_string()),
+        )
+    })?;
+    Ok(lf)
+}
+
+/*
```

Comment:
Minor: for extra points, could make this a proper docstring :)

Reply:
Done.